src/enforcers/MultiTokenPeriodEnforcer.sol

```diff
@@ -0,0 +1,302 @@
+    }
+
+    /**
+     * @notice Decodes all configurations contained in _terms.
+     * @dev Expects _terms length to be a multiple of 116.
+     * @param _terms A concatenation of 116-byte configurations.
+     * @return tokens_ An array of token addresses.
+     * @return periodAmounts_ An array of period amounts.
+     * @return periodDurations_ An array of period durations (in seconds).
+     * @return startDates_ An array of start dates for the first period.
+     */
+    function getAllTermsInfo(bytes calldata _terms)
+        public
+        pure
+        returns (
+            address[] memory tokens_,
+            uint256[] memory periodAmounts_,
+            uint256[] memory periodDurations_,
+            uint256[] memory startDates_
+        )
+    {
+        uint256 termsLength_ = _terms.length;
+        require(termsLength_ % 116 == 0 && termsLength_ != 0, "MultiTokenPeriodEnforcer:invalid-terms-length");
+        uint256 numConfigs_ = termsLength_ / 116;
+        tokens_ = new address[](numConfigs_);
+        periodAmounts_ = new uint256[](numConfigs_);
+        periodDurations_ = new uint256[](numConfigs_);
+        startDates_ = new uint256[](numConfigs_);
+
+        for (uint256 i = 0; i < numConfigs_; i++) {
+            uint256 offset_ = i * 116;
+            tokens_[i] = address(bytes20(_terms[offset_:offset_ + 20]));
+            periodAmounts_[i] = uint256(bytes32(_terms[offset_ + 20:offset_ + 52]));
+            periodDurations_[i] = uint256(bytes32(_terms[offset_ + 52:offset_ + 84]));
+            startDates_[i] = uint256(bytes32(_terms[offset_ + 84:offset_ + 116]));
+        }
+    }
+
+    ////////////////////////////// Internal Methods //////////////////////////////
+
+    /**
+     * @notice Validates and consumes a transfer (native or ERC20) by ensuring the amount does not exceed the available limit.
+     * @dev Decodes the execution data based on token type:
+     *      - For native transfers (_token == address(0)): decodes (target, value, callData) and requires callData to be empty.
+     *      - For ERC20 transfers (_token != address(0)): decodes (target, , callData) and requires callData length to be 68 with a
+     * valid IERC20.transfer selector.
+     * @param _terms The concatenated configurations.
+     * @param _executionCallData The encoded execution data.
+     * @param _delegationHash The delegation hash.
+     * @param _redeemer The address intended to receive the tokens/ETH.
+     */
+    function _validateAndConsumeTransfer(
+        bytes calldata _terms,
+        bytes calldata _executionCallData,
+        bytes32 _delegationHash,
+        address _redeemer
+    )
+        private
+    {
+        uint256 transferAmount_;
+        address token_;
+
+        // Decode _executionCallData using decodeSingle.
+        (address target_, uint256 value_, bytes calldata callData_) = _executionCallData.decodeSingle();
+        if (value_ > 0) {
```

Comment:
even though this is probably fine. Feels off. Maybe we should be switching on logic if it is value == 0 && callData_.length == 68 ?

Reply:
so maybe instead we do
```
if (callData_.length == 68){
  if (value_ > 0) revert;
  // ERC20 transfer
} else {
  if (value_ == 0) revert;
  // Native transfer
  
}
```